server/src/main/java/org/opensearch/action/search/SearchResponse.java

```diff
@@ -91,6 +93,8 @@ public class SearchResponse extends ActionResponse implements StatusToXContentOb
     private final Clusters clusters;
     private final long tookInMillis;
 
+    private List<SearchExtBuilder> searchExtBuilders = new ArrayList<>();
```

Comment:
Having spent some time recently fighting with constructing new `SearchResponse` instances, it seems like a conscious decision was made at some point to make `SearchResponse` immutable, so adding a mutable field feels (potentially) wrong.

Also, I'm not entirely sure if `SearchResponse` is the right place to put this or if it should be added to `SearchResponseSections`. It looks like `SearchResponseSections` is where most "real response stuff" goes, with the `writeTo` serialization handled by `InternalSearchResponse`.

Reply:
I would then suggest that we take a safer route of honoring the immutability and introduce a new ctor that takes in a `List<SearchExtBuilder>` (in addition to all the existing arguments).  Each `SearchResponseProcessor` will be responsible for returning a new `SearchResponse` object with whatever `SearchExtBuilder`s it wants to tack on, on the way out.

Regarding `SearchResponse` vs `SearchResponseSections`, based on my read of the codebase, the former seems like the "outer" layer and the latter, the "inner" layer of what constitutes a search response.  So, I think something like "ext" (extra stuff) feels like it should belong in the outer layer.

Reply:
@msfroh @austintlee I would strongly -1 the approach of adding the the "bag of anything" to the search response, however the could be the cases when search processors may add some contextual information into the response. The way I would have approached that:
 - add the `search_pipeline` parameter to response (mirroring the `search_pipeline`)
 - add per-processor contextual information (some like <processor>: { key1 : value1, ... }

That would make a few things clear:
 - search pipeline was applied
 - processor was run over and provided some data

Reply:
I am not proposing to add arbitrary stuff to SearchResponse, though.

It specifically mirrors the association between search pipelines and SearchRequest:

https://github.com/opensearch-project/OpenSearch/blob/6a5b464c71964800507d389cfb0ae2d358175857/server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java#L205

Reply:
> I am not proposing to add arbitrary stuff to SearchResponse, though.

This is what it looks like to user: the `ext` JSON field with arbitrary data

> It specifically mirrors the association between search pipelines and SearchRequest:

💯 and so we should be designing it as such

Reply:
> It specifically mirrors the association between search pipelines and SearchRequest:

Note that the `SearchExtBuilder` has nothing to do with search pipelines historically. It was added to `SearchSourceBuilder` almost 7 years ago: https://github.com/opensearch-project/OpenSearch/commit/90ab460fccd52db2fa98890cd1268538f57b1c93#diff-dcc359a9e84ab231711e9f30d1989b64e60d27dc229c141a9cc2b9f45cde8d2f

Reply:
Oh thanks @msfroh , @austintlee my apologies for the confusion, from description I concluded that the `ext` part is related to search processors (which in some form were implemented as part of search pipeline).

Reply:
So, what do you guys think?  Should I go with the ctor approach I suggested above?  Do we want to ping more folks to comment on this?  Feel free to tag people you think should weigh in.  Thanks!

Reply:
Circling back to the beginning, I think that 

> Also, I'm not entirely sure if SearchResponse is the right place to put this or if it should be added to SearchResponseSections.

This seems to be the right place since this is indeed a new section of the response. However, since we do have proper search pipelines (and this particular feature request is related to adding some data from search processors to search response), why should we use the `ext` part in search response?

Reply:
@reta Here's an example of how this might look:
```
{
  "took": 3,
  "timed_out": false,
  "_shards": {
    "total": 3,
    "successful": 3,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 110,
      "relation": "eq"
    },
    "max_score": 0.55129033,
    "hits": [
      {
       "_index": "...",
        "_id": "...",
        "_score": 0.55129033,
        "_source": {
          "text": "...",
          "title": "..."
        }
      },
      {
      ...
      }
      ...
      {
      ...
      },
      "ext": {
        "generative_qa": {
          "answer": "..."
    }
  }
} 
```

Reply:
> @reta Here's an example of how this might look:

Thanks @austintlee , this is reasonable, and it is not bad at all. One of the issues we've had in the past is that some features lack context, like for example the `ext` in the response: yes, it does return some data, but what is the context?  injected by the some plugin? extension? search pipeline response processor?

Reply:
Furthermore, `SearchExtBuild` implements `NamedWritable`:

https://github.com/opensearch-project/OpenSearch/blob/914544b182744d7c523f5d1f4f82ab7df256f424/server/src/main/java/org/opensearch/search/SearchExtBuilder.java#L61

https://github.com/opensearch-project/OpenSearch/blob/914544b182744d7c523f5d1f4f82ab7df256f424/libs/core/src/main/java/org/opensearch/core/common/io/stream/NamedWriteable.java#L47

which means that each of the blocks inside `ext` is a named block that the caller can recognize since it should match elements in the _search query.

Reply:
> what is the context? injected by the some plugin? extension? search pipeline response processor?

Yes 😁 

The (intended?) convention that I've been able to piece together from the equivalent in `SearchRequest` is that specifically there needs to be one or more `SearchPlugin`s installed that define the set of available `ext` children. If you specify an `ext` child in a search request and don't have a plugin that knows how to parse it, the coordinator will throw an exception.

I guess that does suggest a bit of risk when doing the same on the response side, since we're assuming the client will either recognize and understand an `ext` child from the server or will ignore it. (IMO a well-behaved client should ignore an `ext` child that it doesn't understand.)

Reply:
> I guess that does suggest a bit of risk when doing the same on the response side, since we're assuming the client will either recognize and understand an ext child from the server or will ignore it

Correct, since we adding new element to the response, the change should be fine (we did that quite a few times already). So we will go with the `ext` response section, right?

Reply:
Can you guys take a look at the latest diff?  If it looks good, I'll add some YAML tests and update the changelog.  We should also update the Search Pipeline doc page to mention this addition.

Reply:
> @msfroh I think `SearchResponseSections` should be the place, or I misunderstood your suggestion?

I'm a little fuzzy on the dividing line between `SearchResponse` and `SearchResponseSections`. As @austintlee said above:

> Regarding `SearchResponse` vs `SearchResponseSections`, based on my read of the codebase, the former seems like the "outer" layer and the latter, the "inner" layer of what constitutes a search response. So, I think something like "ext" (extra stuff) feels like it should belong in the outer layer.

I think that's a reasonable argument. Another way of looking at it is to consider what currently appears where:

### `SearchResponseSections`
* SearchHits
* Aggregations
* Suggest
* timedOut / terminatedEarly
* SearchProfileShardResults
* numReducePhases

### `SearchResponse`
* scrollId
* pointInTimeId
* totalShards / successfulShards / skippedShards
* shardFailures
* clusters
* tookInMillis

Which one is a better fit for `SearchExt`s? I think they would be more at home in `SearchResponseSections`, because they "feel" more like suggestions or aggregations to me (i.e. structured complex info), but I could be convinced otherwise.

Reply:
I don't feel too strongly about either option.  I am not a committer/maintainer so I'll leave the decision to you guys, but we should decide and move on, especially if the difference is minor and only cosmetic.

Reply:
> Which one is a better fit for SearchExts? I think they would be more at home in SearchResponseSections, because they "feel" more like suggestions or aggregations to me (i.e. structured complex info), but I could be convinced otherwise.

:+1 to `SearchResponseSections`, the `ext` is linked with the that particular search response

Reply:
Please, take a look at the latest rev.